src/stash-debug/src/main.rs

```diff
@@ -0,0 +1,282 @@
+                ($col.name(), serde_json::to_value($col.iter(stash).await?)?)
+            };
+        }
+
+        let data = BTreeMap::from_iter(match self {
+            Usage::Catalog => vec![
+                dump_col!(catalog::COLLECTION_CONFIG),
+                dump_col!(catalog::COLLECTION_ID_ALLOC),
+                dump_col!(catalog::COLLECTION_SYSTEM_GID_MAPPING),
+                dump_col!(catalog::COLLECTION_COMPUTE_INSTANCES),
+                dump_col!(catalog::COLLECTION_COMPUTE_INTROSPECTION_SOURCE_INDEX),
+                dump_col!(catalog::COLLECTION_COMPUTE_REPLICAS),
+                dump_col!(catalog::COLLECTION_DATABASE),
+                dump_col!(catalog::COLLECTION_SCHEMA),
+                dump_col!(catalog::COLLECTION_ITEM),
+                dump_col!(catalog::COLLECTION_ROLE),
+                dump_col!(catalog::COLLECTION_TIMESTAMP),
+                dump_col!(catalog::COLLECTION_SYSTEM_CONFIGURATION),
+                dump_col!(catalog::COLLECTION_AUDIT_LOG),
+                dump_col!(catalog::COLLECTION_STORAGE_USAGE),
+            ],
+            Usage::Storage => {
+                let mut data = vec![dump_col!(storage::METADATA_COLLECTION)];
+                // This collection sometimes hasn't been created yet.
+                if stash
+                    .collections()
+                    .await?
+                    .contains(storage::METADATA_EXPORT.name())
+                {
+                    data.push(dump_col!(storage::METADATA_EXPORT));
+                }
+                data
+            }
+        });
+        let data_names = BTreeSet::from_iter(data.keys().map(|k| k.to_string()));
+        if data_names != self.names() {
+            // This is useful to know because it can either be fine (collection
+            // not yet created) or a programming error where this file was not
+            // updated after adding a collection.
+            eprintln!(
+                "unexpected names, verify this program knows about all collections: got {:?}, expected {:?}",
+                data_names,
+                self.names()
+            );
+        }
+        Ok(data)
+    }
+
+    async fn edit(
+        &self,
+        stash: &mut impl Append,
+        collection: String,
+        key: serde_json::Value,
+        value: serde_json::Value,
+    ) -> Result<Option<serde_json::Value>, anyhow::Error> {
+        macro_rules! edit_col {
+            ($col:expr) => {
+                if collection == $col.name() {
+                    let key = serde_json::from_value(key)?;
+                    let value = serde_json::from_value(value)?;
+                    return Ok($col
+                        .upsert_key(stash, &key, &value)
+                        .await?
+                        .map(|v| serde_json::to_value(v).unwrap()));
+                }
+            };
+        }
+
+        match self {
+            Usage::Catalog => {
+                edit_col!(catalog::COLLECTION_CONFIG);
```

Comment:
Is it possible for `names`, `edit`, and `dump` to all share a vec of `TypedCollection`s instead of duplicating them?

Reply:
Nope because it's a generic struct and I think also not object safe (there's a lot of rust goo here I don't understand, but I tried to do your suggestion when first making this tool and I'm medium sure it's not [easily] possible).